File: src/test/java/com/scs/ReadNWriteModelsTest.java
package com.scs;

import com.scs.voxlib.VLVoxFile;
import com.scs.voxlib.VLVoxReader;
import com.scs.voxlib.VLVoxWriter;
import org.junit.jupiter.api.Test;

import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

import static java.nio.file.StandardOpenOption.CREATE;
import static java.nio.file.StandardOpenOption.TRUNCATE_EXISTING;
import static org.assertj.core.api.Assertions.assertThat;


public class ReadNWriteModelsTest {

    private VLVoxFile loadVox(String resourcePath) throws IOException {
        var fullPath = getClass().getResource(resourcePath).getPath();
        try (var reader = new VLVoxReader(new FileInputStream(fullPath))) {
            return reader.read();
        }
    }

    private void testModel(String path, int fileVersion, int modelCount, int voxelCount, int materialCount) throws IOException {
        var file = loadVox(path);

        var tempFilePath = Paths.get("target" + path);
        try (
            var tempFileOut = Files.newOutputStream(tempFilePath, CREATE, TRUNCATE_EXISTING);
            var writer = new VLVoxWriter(tempFileOut)
        ) {
            writer.write(file);
        }

        //Reload and test again
        try (var reader = new VLVoxReader(Files.newInputStream(tempFilePath))) {
            file = reader.read();
        }

        assertThat(file.getVersion()).isEqualTo(fileVersion);

        var models = file.getModelInstances();

        assertThat(models)
            .isNotNull()
            .hasSize(modelCount)
            .element(0)
            .isNotNull();

        int voxelSum = 0;
        for (var modelInstance : models) {
            voxelSum += modelInstance.model.getVoxels().length;
        }
        assertThat(voxelSum).isEqualTo(voxelCount);

        // First colour is always black
        int[] palette = file.getPalette();
        assertThat(palette[0]).isEqualTo(0x00000000);

        for (int i = 1; i < palette.length; i++) {
            assertThat(palette[i]).isNotEqualTo(0x00000000);
        }

        assertThat(file.getMaterials()).hasSize(materialCount);
    }

    @Test
    public void testChrKnight() throws IOException {
        testModel("/chr_knight.vox", 150, 1, 398, 256);
    }

    @Test
    public void testTeapot() throws IOException {
        testModel("/teapot.vox", 150, 1, 28411, 256);
    }

    @Test
    public void testMonu2() throws IOException {
        testModel("/monu2.vox", 150, 1, 150764, 256);
    }
}

File: src/test/kotlin/fr/chsfleury/kvox/impl/ReadNWriteModelsTest.kt
package fr.chsfleury.kvox.impl

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.io.FileInputStream
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.StandardOpenOption

class ReadNWriteModelsTest {

    @Throws(IOException::class)
    private fun loadVox(resourcePath: String): DefaultVoxFile? {
        return javaClass.getResource(resourcePath)
            ?.path
            ?.let {
                DefaultVoxReader(FileInputStream(it))
                    .use { reader -> reader.read() }
            }
    }

    @Throws(IOException::class)
    private fun testModel(path: String, fileVersion: Int, modelCount: Int, voxelCount: Int, materialCount: Int) {
        var file = loadVox(path)!!
        val tempFilePath = Paths.get("target$path")
        Files.newOutputStream(tempFilePath, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING).use { tempFileOut ->
            DefaultVoxWriter(tempFileOut).use { writer ->
                writer.write(
                    file
                )
            }
        }
        DefaultVoxReader(Files.newInputStream(tempFilePath)).use { reader -> file = reader.read() }
        assertThat(file.version).isEqualTo(fileVersion)
        val models = file.modelInstances
        assertThat(models)
            .isNotNull()
            .hasSize(modelCount)
            .element(0)
            .isNotNull()
        var voxelSum = 0
        for (modelInstance in models) {
            voxelSum += modelInstance.model.voxels.size
        }
        assertThat(voxelSum).isEqualTo(voxelCount)

        // First colour is always black
        val palette = file.palette
        assertThat(palette[0]).isEqualTo(0x00000000)
        for (i in 1 until palette.size) {
            assertThat(palette[i]).isNotEqualTo(0x00000000)
        }
        assertThat(file.materials).hasSize(materialCount)
    }

    @Test
    @Throws(IOException::class)
    fun testChrKnight() {
        testModel("/chr_knight.vox", 150, 1, 398, 256)
    }

    @Test
    @Throws(IOException::class)
    fun testTeapot() {
        testModel("/teapot.vox", 150, 1, 28411, 256)
    }

    @Test
    @Throws(IOException::class)
    fun testMonu2() {
        testModel("/monu2.vox", 150, 1, 150764, 256)
    }

}